src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl

```diff
@@ -0,0 +1,195 @@
+
+const AbstractVerticalSpacingGrid = Union{AbstractVerticalSpacingUnderlyingGrid, AbstractVerticalSpacingImmersedGrid} 
+
+#####
+##### Original grid
+#####
+
+retrieve_static_grid(grid) = grid
+
+function retrieve_static_grid(grid::AbstractVerticalSpacingImmersedGrid) 
+    underlying_grid  = retrieve_static_grid(grid.underlying_grid)
+    active_cells_map = !isnothing(grid.interior_active_cells)
+    return ImmersedBoundaryGrid(underlying_grid, grid.immersed_boundary; active_cells_map)
+end
+
+reference_Δzᵃᵃᶠ(grid) = grid.Δzᵃᵃᶠ
+reference_Δzᵃᵃᶜ(grid) = grid.Δzᵃᵃᶜ
+
+function retrieve_static_grid(grid::AbstractVerticalSpacingUnderlyingGrid) 
+
+    Δzᵃᵃᶠ = reference_Δzᵃᵃᶠ(grid)
+    Δzᵃᵃᶜ = reference_Δzᵃᵃᶜ(grid)
+
+    TX, TY, TZ = topology(grid)
+
+    args = []
+    for prop in propertynames(grid)
+        if prop == :Δzᵃᵃᶠ
+            push!(args, Δzᵃᵃᶠ)
+        elseif prop == :Δzᵃᵃᶜ
+            push!(args, Δzᵃᵃᶜ)
+        else
+            push!(args, getproperty(grid, prop))
+        end
+    end
+
+    GridType = getnamewrapper(grid)
+
+    return GridType{TX, TY, TZ}(args...)
+end
+
+#####
+##### Some extensions
+#####
+
+function with_halo(new_halo, grid::AbstractVerticalSpacingUnderlyingGrid)
+    old_static_grid = retrieve_static_grid(grid)
+    new_static_grid = with_halo(new_halo, old_static_grid)
+    vertical_coordinate = denomination(grid)
+    new_grid = generalized_spacing_grid(new_static_grid, vertical_coordinate)
+
+    return new_grid
+end
+
+function with_halo(new_halo, ibg::AbstractVerticalSpacingImmersedGrid) 
+    underlying_grid = ibg.underlying_grid
+    immersed_boundary = ibg.immersed_boundary
+    new_underlying_grid = with_halo(new_halo, underlying_grid)
+    active_cells_map = isnothing(ibg.interior_active_cells)
+    new_ibg = ImmersedBoundaryGrid(new_underlying_grid, immersed_boundary; active_cells_map)
+    return new_ibg
+end
+
+#####
+##### General implementation
+#####
+
+generalized_spacing_grid(grid, coord) = grid
+update_vertical_spacing!(model, grid; kwargs...) = nothing
+
+##### 
+##### Vertical spacings for a generalized vertical coordinate system
+#####
+
+# Very bad for GPU performance!!! (z-values are not coalesced in memory for z-derivatives anymore)
+# TODO: make z-direction local in memory by not using Fields
+# TODO: make it work with partial cells 
+
+@inline Δzᶜᶠᶠ(i, j, k, grid::AbstractVerticalSpacingGrid) = ℑyᵃᶠᵃ(i, j, k, grid, Δzᶜᶜᶠ)
+@inline Δzᶜᶠᶜ(i, j, k, grid::AbstractVerticalSpacingGrid) = ℑyᵃᶠᵃ(i, j, k, grid, Δzᶜᶜᶜ)
+
+@inline Δzᶠᶜᶠ(i, j, k, grid::AbstractVerticalSpacingGrid) = ℑxᶠᵃᵃ(i, j, k, grid, Δzᶜᶜᶠ)
+@inline Δzᶠᶜᶜ(i, j, k, grid::AbstractVerticalSpacingGrid) = ℑxᶠᵃᵃ(i, j, k, grid, Δzᶜᶜᶜ)
+
+@inline Δzᶠᶠᶠ(i, j, k, grid::AbstractVerticalSpacingGrid) = ℑxyᶠᶠᵃ(i, j, k, grid, Δzᶜᶜᶠ)
+@inline Δzᶠᶠᶜ(i, j, k, grid::AbstractVerticalSpacingGrid) = ℑxyᶠᶠᵃ(i, j, k, grid, Δzᶜᶜᶜ)
+
+@inline Δz_reference(i, j, k, Δz::Number) = Δz
```

Comment:
I'm not a huge fan of this notation because it mixes symbols (`Δz`) with language (`reference`)